### PR TITLE
Sb application

### PR DESCRIPTION
**Pull request by Kjir**
### Short Description
This is my solution to the challenge. My decisions have been described in the SOLUTIONS.md file

### Checks

- [x] This PR follows the coding guidelines of the project
- [x] This PR fulfills the acceptance criteria defined on the issue
- [x] This PR has been locally tested
- [x] Changes in this PR are unit tested

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -29,6 +27,36 @@
...
+				</executions>
+				<configuration>
+					<googleJavaFormatOptions>
+						<aosp>false</aosp>
+						<fixImportsOnly>false</fixImportsOnly>
+						<skipSortingImports>false</skipSortingImports>
+						<skipRemovingUnusedImports>false</skipRemovingUnusedImports>
+					</googleJavaFormatOptions>
+				</configuration>
+			</plugin>
```
Comment (robertblust) [inline pom.xml]: This is cool staff 👍 never used this plugin. Nicely ensures the right formatting.

File: src/main/java/ch/aaap/assignment/raw/CSVUtil.java [hunk 2]
```diff
@@ -26,61 +28,62 @@
...
   }
+
+  public static ApplicationModel getApplicationModel() {
+    return new ApplicationModel(getPoliticalCommunities(), getPostalCommunities());
+  }
+
+  private static <T> Set<T> parseCSV(String csvFile, RowParser<T> rp) throws IOException {
+    InputStream is = CSVUtil.class.getResourceAsStream(csvFile);
+    Reader reader = new InputStreamReader(is, StandardCharsets.UTF_8);
+    @Cleanup CSVParser parser = new CSVParser(reader, CSVFormat.DEFAULT.withHeader());
```
Comment (robertblust) [inline src/main/java/ch/aaap/assignment/raw/CSVUtil.java]: Nice usage of lombok clean up. And good refactoring to remove duplicated code 🏅

File: src/test/java/ch/aaap/assignment/ApplicationTest.java [hunk 3]
```diff
@@ -80,6 +78,20 @@ public void returnsCorrectDistrictNameForZipCode() {
         sut.getDistrictForZipCode("7457"),
         "Correct district name for zip code '7457'");
 
+    assertEquals(
+        "Wahlkreis See-Gaster",
+        sut.getDistrictForZipCode("8866"),
+        "Correct district name for zip code '8866'");
+
+    // FIXME: The following test will fail because it is not possible
```
Comment (robertblust) [inline src/test/java/ch/aaap/assignment/ApplicationTest.java]: That's an excellent finding! CH zip codes looks like such a simple Domain from outside, but it's quite tricky. My current thinking was that a zip code to district is a 1 to 1 relation. But you just invalidated my thinking. Funny that in our reference solution the test case fails for 'Wahlkreis See-Gaster' and works for 'Kanton Glarus'. Our application model looks slightly different. 

Good catch 👍How you found this? You have some relations to zip code '8866'? I guess I need to implement some more analytics to see how many zip code belong to more then one district.

File: SOLUTION.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
+# Political and Postal Communities of Switzerland
+## Solution
+
+This is my solution to the coding challenge. I took some decisions that I'd like to explain here to give a little bit of context.
+
+### git history
+
+The git history is relevant. The commits are organized meaningfully and show my incremental approach to solving the problem. Some steps are implementation, some steps are refactoring, some others are just reformatting or similar.
+
```
Comment (robertblust) [inline SOLUTION.md]: I like. Sometimes hard to achieve under time pressure, but it really helps a reviewer to understand your solution approach.

File: SOLUTION.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+### Formatting
+I initially tried automatic formatting within my editor, but I soon realised that the setup wasn't working properly. I ultimately added a package that creates a git hook and automatically formats code when committing. Ideally that should have been a first step.
+
```
Comment (robertblust) [inline SOLUTION.md]: 👍 always check on the gate makes totally sense. At 3AP, the IDE is free of choice and what we want to have is clean code on git at the end.

File: SOLUTION.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+### Refactoring
+I refactored the CSV parsing code to make it more functional and make it more testable. While not strictly necessary, it allowed me to reduce code duplication.
+
```
Comment (robertblust) [inline SOLUTION.md]: 👍

File: SOLUTION.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+### Testing
+I didn't add any new tests on top of those existing to guide the solution. However the test coverage is probably not optimal and some more unit tests are probably warranted. I decided not to focus on that aspect for this excercise because I'm not too familiar with jUnit and it would have required a larger amount of time and I would rather have a working solution first, before refactoring the tests to be more comprehensive.
+
+### Interfaces to classes
+I decided to transform the given interfaces to classes. I thought about this and I felt uncomfortable with having the interface and a `CantonImpl` or `DefaultCanton` class to implement that interface because I do not see it very likely that there will ever be a different implementation for those interfaces. If my assumptions turn out to be wrong, it is an easy refactor.
+
```
Comment (robertblust) [inline SOLUTION.md]: Totally agree. I like your understanding when an interface is required or not. The interface are just there to give a hint the developers in this case. Pragmatic solutions are welcome at 3AP.

File: SOLUTION.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+### Builder pattern
+I avoided the builder pattern because I felt it wasn't necessary. We always have all the parameters we need and the builder pattern doesn't add anything to what we have right now. This might change with the evolution of the software, but I always try to abide by the YAGNI pricinple.
+
```
Comment (robertblust) [inline SOLUTION.md]: don't implement something until it is necessary 😘

File: SOLUTION.md [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+### Memoization
+I chose to prefer data transformations to having objects with arbitrarily long dependency chains. I am trading off a simpler representation at the expense of some computation.
+
+While a more structured data could require less computation for some of the answers, it might make the creation of such objects more complicated, and it could lead to incosistent state in different objects.
+
+A possible mitigation for the computations would be to memoize the results of some functions, so that they don't have to be repeated. However, the current data size and usage do not justify such an effort.
+
+### Bugs
+There is a bug in the original program. I added a commented out test to highlight this bug, which is a problem in the requirements: it is not possible to get a unique district name out of a ZIP code because the function is not reversible (a.k.a. multivalued function, or non-bijective function): there are some ZIP codes that belong to more than one district. An example of such a ZIP code is `8866` which is both in Canton Glarus and St. Gallen.
```
Comment (robertblust) [inline SOLUTION.md]: See my comment in the code. Perfect catch. I would not declare it as bug. I declare it as we gain more domain knowledge.

File: src/test/java/ch/aaap/assignment/ApplicationTest.java (hunk 3 above)
Comment (Kjir) [reply src/test/java/ch/aaap/assignment/ApplicationTest.java]: I just got a hunch and wanted to verify it: is it really true that ZIP codes only apply to one district? I could have worked with the CSV files, but I decided to first try and see if something came up on the web.
Wikipedia had the answer: https://en.wikipedia.org/wiki/Postal_codes_in_Switzerland_and_Liechtenstein

**Pull request closed without merging**